borg/archiver.py

```diff
@@ -54,20 +55,69 @@ def __call__(self, parser, ns, values, option):
         setattr(ns, self.dest, not option.startswith('--no-'))
 
 
+def argument(args, s):
+    if isinstance(s, str):
+        return getattr(args, s)
+    return s
+
+
+def drop_unused_args(function):
+    @functools.wraps(function)
+    def wrapper(*args, **kwargs):
+        function_args = list(signature(function).parameters)
+        for key in list(kwargs):
+            if key not in function_args:
+                del kwargs[key]
+        if len(kwargs) + len(args) > len(function_args):
+            args = args[:len(function_args) - len(kwargs)]
+        return function(*args, **kwargs)
+    return wrapper
+
+
+def with_repository(fake=False, create=False, exclusive=False, lock=True, manifest=True):
+    def decorator(method):
+        @functools.wraps(method)
+        def wrapper(self, args, **kwargs):
+            location = args.location  # note: 'location' must be always present in args
+            if argument(args, fake):
+                return method(self, args, repository=None, **kwargs)
+            elif location.proto == 'ssh':
+                repository = RemoteRepository(location, create=create, lock_wait=self.lock_wait, lock=lock, args=args)
+            else:
+                repository = Repository(location.path, create=create, exclusive=argument(args, exclusive),
+                                        lock_wait=self.lock_wait, lock=lock)
+            with repository:
+                if manifest and not argument(args, fake):
+                    kwargs['manifest'], kwargs['key'] = Manifest.load(repository)
+                return method(self, args, repository=repository, **kwargs)
+        return wrapper
+    return decorator
+
+
+def with_archive(method):
+    @functools.wraps(method)
+    def wrapper(self, args, repository, key, manifest, **kwargs):
+        archive = Archive(repository, key, manifest, args.location.archive,
+                          numeric_owner=getattr(args, 'numeric_owner', False), cache=kwargs.get('cache'))
+        return method(self, args, repository=repository, manifest=manifest, key=key, archive=archive, **kwargs)
+    return wrapper
+
+
+def with_cache(method):
+    @functools.wraps(method)
+    def wrapper(self, args, repository, key, manifest, **kwargs):
+        with Cache(repository, key, manifest, do_files=getattr(args, 'cache_files', False),
+                   lock_wait=self.lock_wait) as cache:
+            return method(self, args, repository=repository, key=key, manifest=manifest, cache=cache, **kwargs)
+    return wrapper
+
```

Comment:
I don't see much magic in the with_* decorators, but I agree that `@drop_unused_args` might be too opaque to actually use it outside a Python Obfuscated Code Contest (well, not that bad, but you get the idea).